File: esw-ocs/esw-ocs-dsl-kt/src/main/kotlin/esw/ocs/dsl/highlevel/LoopDsl.kt
package esw.ocs.dsl.highlevel

import kotlinx.coroutines.*
import kotlin.time.Duration
import kotlin.time.milliseconds

interface LoopDsl {
    companion object {
        private val loopInterval: Duration = 50.milliseconds
    }

    val coroutineScope: CoroutineScope

    /****** Sequential loops *******/
    suspend fun loop(block: suspend StopWhen.() -> Unit): Job = loop(loopInterval, block)

    suspend fun loop(minInterval: Duration, block: suspend StopWhen.() -> Unit): Job = loop0(minInterval, block)

    /****** Background loops *******/
    fun bgLoop(block: suspend StopWhen.() -> Unit): Deferred<Job> = coroutineScope.async { loop(loopInterval, block) }

    fun bgLoop(minInterval: Duration, block: suspend StopWhen.() -> Unit): Deferred<Job> =
        coroutineScope.async { loop(minInterval, block) }

    /****** Waiting for condition to be true *******/
    suspend fun waitFor(condition: suspend () -> Boolean) = loop { stopWhen(condition()) }

    // ========== INTERNAL ===========
    private suspend fun loop0(minInterval: Duration, block: suspend StopWhen.() -> Unit) = coroutineScope {
        launch {
            suspend fun go() {
                delayedResult(maxOf(minInterval, loopInterval), block)
                go()
            }

            go()
        }
    }

    private suspend fun <T> delayedResult(minDelay: Duration, block: suspend StopWhen.() -> T): T = coroutineScope {
        val futureValue = async { block(StopWhen) }
        delay(minDelay.toLongMilliseconds())
        futureValue.await()
    }

    object StopWhen {
        suspend fun stopWhen(condition: Boolean): Unit = coroutineScope {
            suspendCancellableCoroutine<Unit> {
                if (condition) it.cancel() else it.resumeWith(Result.success(Unit))
            }
        }
    }
}

File: esw-ocs/esw-ocs-dsl-scala/src/main/scala/esw/ocs/dsl2/highlevel/LoopDsl.scala
package esw.ocs.dsl2.highlevel

import esw.ocs.dsl.script.StrandEc
import esw.ocs.dsl2.highlevel.LoopDsl.{ExitLoop, loopInterval}

import java.util.concurrent.ScheduledExecutorService
import scala.concurrent.duration.{Duration, DurationInt}
import scala.concurrent.{ExecutionContext, Future, Promise}
import async.Async.*
import scala.annotation.tailrec

class LoopDsl(strandEc: StrandEc) {
  given ExecutionContext                 = strandEc.ec
  def break(): Nothing                   = throw ExitLoop()
  def stopWhen(condition: Boolean): Unit = if (condition) break()

  inline def loop(inline block: Unit): Unit                        = loop(loopInterval)(block)
  inline def loop(minInterval: Duration)(inline block: Unit): Unit = await(loopAsync(minInterval)(block))
  inline def waitFor(condition: Boolean): Unit                     = loop(stopWhen(condition))

  inline def loopAsync(inline block: Unit): Future[Unit]                        = loopAsync(loopInterval)(block)
  inline def loopAsync(minInterval: Duration)(inline block: Unit): Future[Unit] = loop0(minInterval)(async(block))

  private def loop0(minInterval: Duration)(block: => Future[Unit]): Future[Unit] = async {
    try
      await(delayedResult(minInterval.max(loopInterval))(block))
      await(loop0(minInterval)(block))
    catch case ex: ExitLoop => ()
  }

  private def delayedResult[T](minDelay: Duration)(block: => Future[T]): Future[T] = async {
    delay(minDelay)
    await(block)
  }

  inline def delay(duration: Duration): Unit = await(delayAsync(duration))

  private def delayAsync(duration: Duration): Future[Unit] =
    val p: Promise[Unit] = Promise()

    strandEc.executorService.schedule(
      () => p.trySuccess(()),
      duration.length,
      duration.unit
    )

    p.future
  end delayAsync

}

object LoopDsl {
  private val loopInterval: Duration = 50.milliseconds
  private class ExitLoop() extends RuntimeException
}
